pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -738,6 +639,46 @@ func (c *Controller) getProxyServiceInstancesByEndpoint(endpoints v1.Endpoints,
 	return out
 }
 
+func (c *Controller) getProxyServiceInstancesByEndpointSlice(slice discoveryv1alpha1.EndpointSlice, proxy *model.Proxy) []*model.ServiceInstance {
+	out := make([]*model.ServiceInstance, 0)
+
+	hostname := kube.ServiceHostname(slice.Labels[discoveryv1alpha1.LabelServiceName], slice.Namespace, c.domainSuffix)
+	c.RLock()
+	svc := c.servicesMap[hostname]
+	c.RUnlock()
+
+	if svc != nil {
+		for _, port := range slice.Ports {
+			if port.Name == nil || port.Port == nil {
+				continue
+			}
+			svcPort, exists := svc.Ports.Get(*port.Name)
+			if !exists {
+				continue
+			}
+
+			podIP := proxy.IPAddresses[0]
+
+			// consider multiple IP scenarios
+			for _, ip := range proxy.IPAddresses {
+				for _, ep := range slice.Endpoints {
+					for _, a := range ep.Addresses {
```

Comment:
There are three tiers of loop, can we wrap a function to find the EndpointSlice.Endpoint where the proxy belongs.   And i think this query is O(N) by proxy.IPAddresses[0],  and then it is simple

Reply:
I can break it out but I'm not worried about the performance of the 3 loops. Most of the time there is 1 or 2 ip, 1 or 2 address, and at most 100 endpoints

Reply:
In most cases, it is true. But we should not take users that require high scalability and also may have multi-network devices per pod into consideration.